Add database migration test

rvl opened the pull request:
Relates to #1177.

# Overview

Adds a migration test that can be run against different versions of the wallet.

The idea is to use an api client to set up the database on a wallet server running one version. Then start up the wallet server on the current version and use the api client to check the database.

All the possible upgrade paths from previous released versions to the current version _could_ be exhaustively tested. However, in this initial version, only the direct upgrade from each release to latest is tested.

At present the testing actions are quite basic. It first restores a wallet, migrates, then lists the wallets.

The migration tests are run as part of nightly CI.

rvl committed:
Add database migration test

Migration test uses untyped requests which means that it is most
likely to be compatible with a wide range of versions.

There is a zip file in release.nix for the Windows build of these
tests.

KtorZ commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -104,6 +105,7 @@ library
       Cardano.Pool.Metadata
       Cardano.Wallet
       Cardano.Wallet.Api
+      Cardano.Wallet.Api.Client
```
:+1:

KtorZ commented on nix/migration-tests.nix:
```diff
@@ -0,0 +1,105 @@
+      allowFail = true; }
+    { rev = "v2019-12-16";
+      sha256 = "0y5xf43lrc7gygvxh5ywkglr3d1xcc19dsskm7frl0v6m9yxzni6"; }
+  ];
```
:thinking: I am not sure we have to go that far in the past. In particular, `v2019-11-18` is certainly not compatible with `v2019-12-09`, isn't it ?

KtorZ commented on lib/jormungandr/test/migration/migration-test.hs:
```diff
@@ -0,0 +1,236 @@
+    wal <- post (url apiBase "wallets") (toJSON postData)
+    let walId = wal ^. responseBody . key "id" . _String
+    logInfo tr $ "Create wallet with id " <> walId
+    void $ waitForRestore apiBase walId
```
**[question]**

What is the need for this call? It doesn't seem that the state of the wallet would have any influence on the step2 :thinking: ?

rvl replied:
There is no specific need, but generally, migrations are more likely to fail if the DB tables have some rows in them. That's why it creates a wallet and waits for it to restore.

When we actually implement some migrations, we should add extra test data, but that would depend on what the migrations are.

KtorZ commented on nix/launch-migration-test.sh:
```diff
@@ -0,0 +1,27 @@
+  exit 1
+fi
+
+: "${src:=$(cd `dirname $0`/..; pwd)}"
```
Slightly magic at first glance. Had to look up the semantic of the colon `:` which actually still feels like a magic trick (though I guess most bash scripts are in a way a condensed collection of magic tricks :upside_down_face: )

Why not simply ```src=$(cd `dirname $0`/..; pwd)``` ? Do we expect `src` to be possibly defined globally ?

rvl replied:
The `src` environment variable that can be set externally so that the script can be used both inside and outside of a nix build.

I thought the `:` for variable default values was standard well-known bash awfulness.

KtorZ replied:
I was not referring to the `:` in `src:=`, but to the leading one. First time I see this of my life to be honest :sweat_smile: ... and I've written quite some bash already.

KtorZ reviewed (commented):
Nice work with this. I am worried with one thing though, I'd expect this to run into: https://github.com/input-output-hk/cardano-wallet/issues/1251 but it doesn't :thinking: ...

rvl committed:
Rework migration test to use correct build for step2

And it adds more assertions.

KtorZ committed:
control genesis hash inside known wallet upon restarting workers

KtorZ committed:
use latest available genesis config for ALL migration tests

Use the genesis block from the latest release only.
Having different genesis block (hash) across releases will basically
make all wallets incompatible with each others. Prior to _now_ workers
would simply loop ad-infinitum trying to rollback to (0, 0).

Now, workers would simply exit complaining about a different genesis
hash. To work around this, we use the same genesis configuration for
all nodes in the migration tests, so that the hash remains the same
across all tests.

KtorZ committed:
fixup! 8c4c43d35 use 'tableName' instead of 'fieldName'

KtorZ committed:
remove redundant dependencies from migration tests
